prog/rand.go

```diff
@@ -754,3 +761,84 @@ func (a *CsumType) generate(r *randGen, s *state) (arg Arg, calls []*Call) {
 	// Filled at runtime by executor.
 	return MakeConstArg(a, 0), nil
 }
+
+// Finds a compatible resource with the type `t` and the calls that initialize that resource.
+func resourceCentric(t *ResourceType, s *state, r *randGen) (resource *ResultArg, calls []*Call) {
+	if len(s.corpus) == 0 {
+		return nil, nil
+	}
+
+	corpusResources := make(map[*ResultArg](map[*Call]bool))
+	for _, p := range s.corpus {
+		addCompatibleResources(p.Clone(), t.TypeName, r, corpusResources)
+	}
+
+	if len(corpusResources) == 0 {
+		return nil, nil
+	}
+
+	idx := r.Intn(len(corpusResources))
+	for res, callsRes := range corpusResources {
+		if idx != 0 {
+			idx--
+			continue
+		}
+		resource = res
+		for c := range callsRes {
+			ForeachArg(c, func(arg Arg, _ *ArgCtx) {
+				a, ok := arg.(*ResultArg)
+				if !ok || resource == a || (a.uses != nil && a.uses[a]) {
+					return
+				}
+				if a.typ.Dir() == DirOut {
+					return
+				}
+				// Replace the resource with a new default argument.
+				replaceArg(a, a.typ.DefaultArg())
+			})
+			calls = append(calls, c)
+		}
+		break
+	}
+
+	return
+}
+
+// Adds to the `corpusResources` all the compatible resources with `resourceType` from a program
+// and the corresponding sequence of calls.
+func addCompatibleResources(p *Prog, resourceType string, r *randGen, corpusResources map[*ResultArg](map[*Call]bool)) {
+	for _, c := range p.Calls {
+		ForeachArg(c, func(arg Arg, _ *ArgCtx) {
```

Comment:
When finding a resource you should probably also make sure that it's an output (`DirOut`) resource.